### PR TITLE
 Add columns to pg_dist_object for distributed functions

### PR DESCRIPTION
**Pull request by onderkalaci**
Both my work and #2965 relies on the changes here. So, let's merge this to master and do not block/rely on each other.

@serprex This change is already based on #2965, so can you please review this? 

Note: The internal data structures always rely on the index (even if we might consider using parameter name), because that's cheaper to keep and process.

### CODE REVIEW

**Comment by serprex** [comment]
@onderkalaci I think having 2 functions makes more sense

**Comment by onderkalaci** [comment]
> @onderkalaci I think having 2 functions makes more sense

After reading my comments again, that also makes more sense to me actually.

File: src/backend/distributed/sql/citus--8.3-1--9.0-1.sql [hunk 1]
```diff
@@ -33,6 +33,11 @@ CREATE TABLE citus.pg_dist_object (
     objid oid NOT NULL,
     objsubid integer NOT NULL,
 
+    -- fields that are only valid for distributed
+    -- functions/procedures
+    distribution_argument_index int,
+    colocationid int,
```
Comment (marcocitus) [inline src/backend/distributed/sql/citus--8.3-1--9.0-1.sql]: It would make a bit more sense for these to be at the end, since we might add more of these class-specific fields later.

File: src/backend/distributed/sql/udfs/create_distributed_function/latest.sql [hunk 2]
```diff
@@ -1,7 +1,12 @@
-CREATE OR REPLACE FUNCTION create_distributed_function(function_name regprocedure)
+CREATE OR REPLACE FUNCTION create_distributed_function(function_name regprocedure,
+						       colocate_with text DEFAULT 'default',
+						       distribution_argument_name text DEFAULT '',
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/create_distributed_function/latest.sql]: Seems fine. I would have expected that the arguments have the same order as create_distributed_table in the happy case, but could be convinced otherwise.

File: src/backend/distributed/sql/udfs/create_distributed_function/latest.sql [hunk 3]
```diff
@@ -1,7 +1,12 @@
-CREATE OR REPLACE FUNCTION create_distributed_function(function_name regprocedure)
+CREATE OR REPLACE FUNCTION create_distributed_function(function_name regprocedure,
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/create_distributed_function/latest.sql]: Should this change be in a separate commit or PR? This doesn't seem relevant to other PRs without an actual implementation.

We should also make sure to use qualified names in these scripts or move them into the upgrade script such that the search path is clear.

Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/create_distributed_function/latest.sql]: Ok, makes sense. We can discuss the API on another PR.

File: src/include/distributed/pg_dist_object.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/*-------------------------------------------------------------------------
```
Comment (pykello) [inline src/include/distributed/pg_dist_object.h]: This file should be merged into src/include/distributed/metadata/pg_dist_object.h. Two issues:
1. Both start with `#ifndef PG_DIST_OBJECT_H #define PG_DIST_OBJECT_H`,
2. Both define `Anum_pg_dist_object...`.
